### PR TITLE
Update git clone logic in build.xml

### PR DESCRIPTION
**Pull request by annaibm**
Removes prestaged jck material and reclones it, if git command fails.

resolves: https://github.com/adoptium/aqa-tests/issues/4846

### CODE REVIEW

**Comment by annaibm** [comment]
Grinder_jck test: 2010/2011

File: jck/build.xml [hunk 1]
```diff
@@ -101,8 +118,13 @@
 						<equals arg1="${return.code}" arg2="0"/>
 					</not>
 					<then>
+						<!-- Git command failed, remove pre-staged material and re-clone -->
 						<echo message="Error: Git command failed with return code ${return.code}. Delete the directory and continue." />
-						<delete dir="${JCK_ROOT_USED}" failonerror="true" />
+						<delete dir="${JCK_ROOT_USED}" failonerror="false" />
```
[the remark this answers is not in the document]
Comment (LongyuZhang) [reply jck/build.xml:123]: Also, if you delete ${JCK_ROOT_USED}, then git clone directly, I think the folder structure will be different than what line 163 is. maybe should delete ${JCK_GIT_REPO_USED} instead?

File: jck/build.xml [hunk 2]
```diff
@@ -139,14 +161,8 @@
...
-				<exec executable="git" dir="${JCK_ROOT_USED}/.." failonerror="true">
-					<arg value="clone" />
-					<arg value="--single-branch" />
-					<arg value="-b"/>
-					<arg value="${jck_branch}"/>
-					<arg value="${JCK_GIT_REPO_USED}" />
-					<arg value="${JCK_ROOT_USED}" />
-				</exec>
+				<!-- Use the git-clone macro -->
+					<git-clone branch="${jck_branch}" repo="${JCK_GIT_REPO_USED}" target-dir="${JCK_ROOT_USED}" />
```
Comment (karianna) [inline jck/build.xml:165]:
```suggestion
				<git-clone branch="${jck_branch}" repo="${JCK_GIT_REPO_USED}" target-dir="${JCK_ROOT_USED}" />
```

**Review by karianna — APPROVED** [review]

**Review by LongyuZhang — APPROVED** [review]
LGTM. Thanks @annaibm

**Review by llxia — CHANGES_REQUESTED** [review]
This problem is not limited to git clone. It is for general git cmd. I think it would be easier to rewrite the logic from scratch. ant script may not be the best way to do it.

**Comment by llxia** [comment]
Converted the PR in draft. Please change the PR in `Ready for review` when it is ready.

File: jck/build.xml [hunk 3]
```diff
@@ -15,14 +15,15 @@
 -->
 
 <project name="jck" default="build" basedir=".">
+	<!-- Define a macro for Git clone with error handling -->
```
Comment (LongyuZhang) [inline jck/build.xml:18]: This comment could be removed

File: jck/build.xml [hunk 4]
```diff
@@ -15,14 +15,15 @@
...
 	<taskdef resource="net/sf/antcontrib/antlib.xml" />
 	<description>Build JCK Tests</description>
 
 	<!-- set global properties for this build -->
 	<property name="DEST" value="${BUILD_ROOT}/jck" />
 	<property environment="env" />
 
-	<condition property="isZOS" value="true">
+	<condition property="isZOS" value="true" else="">
```
Comment (LongyuZhang) [inline jck/build.xml:26]: Setting `else` to `false` would be better to avoid parameter pass in errors.

File: jck/build.xml [hunk 5]
```diff
@@ -87,156 +88,16 @@
...
-		</if> 
+		<exec executable="bash" failonerror="false">
+			<arg value="update_git_material.sh"/>
+			<!-- Add other arguments as needed -->
+			<!-- Pass all properties as environment variables -->
+			<!-- Pass all properties as environment variables -->
+			<env key="JCK_ROOT_USED" value="${JCK_ROOT_USED}" />
+			<env key="JCK_GIT_REPO_USED" value="${JCK_GIT_REPO_USED}" />
+			<env key="jck_branch" value="${jck_branch}" />
+			<env key="isZOS" value="${isZOS}" />
```
Comment (LongyuZhang) [inline jck/build.xml:99]: Setting `env` variables are not necessary here, you can pass in them as parameters of the shell script, you can refer this test for it ([link](https://github.com/adoptium/aqa-tests/blob/c089b7328234947ed6cc7d24632c084049da5e5c/external/criu-portable-checkpoint/build.xml#L21-L27))

File: jck/build.xml [hunk 6]
```diff
@@ -87,156 +88,16 @@
...
-						</exec>
-						<exec executable="git" dir="${JCK_ROOT_USED}" failonerror="false">
-							<arg value="reset" />
-							<arg value="--hard" />
-						</exec>
-					</else>
-				</if>
-			</then>
-		</if> 
+		<exec executable="bash" failonerror="false">
```
Comment (LongyuZhang) [inline jck/build.xml:91]: Not sure if we need to set `true` here?

File: jck/update_git_material.sh [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    if [ $? -eq 0 ]; then
+        # Change into the parent directory
+        cd "$(dirname "$clone_directory")" || exit 1
+        # Clone the repository into the target directory
+        git clone --single-branch -b "${jck_branch}" "${JCK_GIT_REPO_USED}" "$(basename "$clone_directory")"
+        # Check if the clone was successful
+        if [ $? -eq 0 ]; then
+            echo "Repository cloned successfully into $clone_directory"
+        else
+            echo "Failed to clone repository"
```
Comment (LongyuZhang) [inline jck/update_git_material.sh:43]: If git clone fails, we need to exit too, not only print the error message

File: jck/update_git_material.sh [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+        fi
+    else
+        echo "Failed to create parent directory: $(dirname "$clone_directory")"
```
Comment (LongyuZhang) [inline jck/update_git_material.sh:46]: Same here for exit.

File: jck/update_git_material.sh [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+    if [ -d "$JCK_ROOT_USED/.git" ] || git -C "$JCK_ROOT_USED" rev-parse --is-inside-work-tree &>/dev/null; then
+        echo "The directory is a valid Git repository."
+        check_and_handle_updates
+    else
+        echo "The directory is not a valid Git repository."
+        remove_and_clone
+    fi    
+else
+    # JCK materials don't exist, git clone
+    git_clone || remove_and_clone
```
Comment (LongyuZhang) [inline jck/update_git_material.sh:144]: If the directory doesn't exist, `git_clone` once would be enough, no need `remove_and_clone`.

File: jck/build.xml [hunk 10]
```diff
@@ -87,156 +87,16 @@
...
-						</exec>
-					</else>
-				</if>
-			</then>
-		</if> 
+		<exec executable="bash" failonerror="true">
+			<arg value="update_git_material.sh"/>
+			<!-- Add other arguments as needed -->
+			<!-- Pass all properties as environment variables -->
+			<!-- Pass all properties as environment variables -->
```
Comment (karianna) [inline jck/build.xml:94]: Duplicate comment

File: jck/update_git_material.sh [hunk 11]
```diff
@@ -0,0 +1,151 @@
+#!/bin/bash
+# Now you can use these variables in your script
```
Comment (karianna) [inline jck/update_git_material.sh:2]:
```suggestion
# Now you can use these variables in your script

```

File: jck/update_git_material.sh [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+JCK_ROOT_USED="$1"
+JCK_GIT_REPO_USED="$2"
+jck_branch="$3"
+isZOS="$4"
+echo "JCK_ROOT_USED is: $JCK_ROOT_USED"
+echo "JCK_GIT_REPO_USED is: $JCK_GIT_REPO_USED"
+echo "jck_branch is: $jck_branch"
+echo "isZOS is: $isZOS"
+
```
Comment (karianna) [inline jck/update_git_material.sh:11]:
```suggestion
```

File: jck/update_git_material.sh [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+        if [ -n "$local_material_uptodate" ]; then
+            echo "zOS Local JCK materials up-to-date. Skipping hard reset"
+        else
+            # Perform hard reset for z/OS
+            perform_hard_reset_zos
+        fi
+    fi
+}
+
+# Function to perform hard reset for z/OS
```
Comment (karianna) [inline jck/update_git_material.sh:123]: Why is this needed?

File: jck/update_git_material.sh [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+    }
+
+    # Clean untracked files and directories
+    git clean -dfx || {
+        echo "Failed to clean untracked files and directories."
+        remove_and_clone
+    }
+
+    # Pull the latest changes
+    git pull "$JCK_GIT_REPO_USED" "$jck_branch" || {
```
Comment (llxia) [inline jck/update_git_material.sh:77]: I know this logic is inherited, but I think there is no need to do git pull again after git reset. 
https://stackoverflow.com/questions/43037293/what-is-the-difference-between-git-pull-and-git-reset-hard-origin-branch
Could you verify if `git pull` is still needed here?

File: jck/update_git_material.sh [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+
+# Function to remove test material and git clone
+remove_and_clone() {
+    echo "Removing test material and git clone again"
+    rm -rf "$JCK_ROOT_USED"
+    git_clone
+}
+
+# Function to perform hard reset and pull
+perform_hard_reset_and_pull() {
```
Comment (llxia) [inline jck/update_git_material.sh:63]: This should handle both zos and non-zos. We do not need perform_hard_reset_zos to handle zos separately.

**Review by llxia — CHANGES_REQUESTED** [review]

**Comment by llxia** [comment]
The current PR only adds the error check in the places that we think it may fail. This is not sufficient. The better way to do this is to use `try/catch`. In shell, we mimic `try/catch` as follows. 

```
{ # try

  # commands

} || { # catch

}
```
https://stackoverflow.com/questions/22009364/is-there-a-try-catch-command-in-bash

Also, this can help us to write clean code to avoid repeatedly calling the same functions. For example, `remove_and_clone` is called 6 times in the PR.

File: jck/update_git_material.sh (hunk 13 above)
Comment (annaibm) [reply jck/update_git_material.sh:122]: Will need to remove this function as `perform_hard_reset_and_pull` needs to handle both ZOS and non-ZOS.

File: jck/update_git_material.sh (hunk 15 above)
Comment (annaibm) [reply jck/update_git_material.sh:63]: @llxia So I need to move the logic of `perform_hard_reset_zos` into  `perform_hard_reset_and_pull` based on ZOS flag or is it general logic which is already there? As I had inherited the logic from build.xml here, so not sure.

File: jck/update_git_material.sh (hunk 14 above)
Comment (annaibm) [reply jck/update_git_material.sh:77]: Sure will test and let you know

File: jck/update_git_material.sh (hunk 15 above)
Comment (llxia) [reply jck/update_git_material.sh:63]: Based on the ZOS flag, ZOS needs to replace the attribute file.

File: jck/update_git_material.sh [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+    fi
+else
+    # JCK materials don't exist, git clone
+    echo "JCK materials don't exist."
+    git_clone
+fi
+# Catch block
+if [ "$global_error" -eq 1 ]; then
+    handle_error
+fi
```
Comment (llxia) [inline jck/update_git_material.sh:127]: Did the following work?
```
{ 
     check_and_handle_updates
} || {
     git_clone
}
```

File: jck/update_git_material.sh [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+	fi
+}
+
+# Function to perform hard reset and pull
+perform_hard_reset() {
+	echo "Performing hard reset..."
+	# Fetch changes from the remote repository
+	git fetch "$JCK_GIT_REPO_USED"
+	# Reset the local branch to match the remote branch
+     git reset --hard "FETCH_HEAD"
```
Comment (llxia) [inline jck/update_git_material.sh:59]: Please fix the indentation. Please check the rest of PR.

File: jck/update_git_material.sh [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+		git clean -fdx
+		local_material_uptodate="true"
+	fi
+	# If SHA match or update succeeded
+   	echo "Local JCK materials updated"
+}
+
+# Try block
+if [ -d "$JCK_ROOT_USED" ] && { [ -d "$JCK_ROOT_USED/.git" ] || git -C "$JCK_ROOT_USED" rev-parse --is-inside-work-tree &>/dev/null; }; then
+	echo "The directory is a valid Git repository."
```
Comment (llxia) [inline jck/update_git_material.sh:109]: We need to clean all unstaged change.
```
git -C $JCK_ROOT_USED clean -fd
```

Comment (annaibm) [reply jck/update_git_material.sh:109]: Yes we are handling it under `check_and_handle_updates` method, does that suffice?

File: jck/update_git_material.sh (hunk 16 above)
Comment (annaibm) [reply jck/update_git_material.sh:127]: I had tried the above method but it doesnt catch errors happening under functions within `perform_hard_reset` or other places as such. So need to add error handling to check the error code returned in addition to trap statement.

File: jck/update_git_material.sh (hunk 18 above)
Comment (llxia) [reply jck/update_git_material.sh:109]: no, it is not. In some cases, someone can manually change the file on disk without committing. Therefore, the SHA is the same. check_and_handle_updates does not catch this case.
